tests/run/rescue.scala

```diff
@@ -0,0 +1,78 @@
+import scala.util.control.NonFatal
+import scala.util.control.NonLocalReturns._
+
+object lib {
+  inline def (op: => T) rescue[T] (fallback: => T) =
+    try op
+    catch {
+      case NonFatal(_) => fallback    // ReturnThrowable is fatal error, thus ignored
+    }
+
+  inline def (op: => T) rescue[T, E <: Throwable] (fallback: PartialFunction[E, T]) =
+    try op
+    catch {
+      // case ex: ReturnThrowable[_] => throw ex      // bug #7041
+      case ex: E =>
+        // user should never match `ReturnThrowable`, which breaks semantics of non-local return
+        if (fallback.isDefinedAt(ex) && !ex.isInstanceOf[ReturnThrowable[_]]) fallback(ex) else throw ex
```

Comment:
Maybe `E` should b a subtype of `Exception` to avoid catching other `Throwable`s like `ReturnThrowable`.

Reply:
This is open to discussion --- this is the case where users specify handled exceptions explicitly, but indeed `ReturnThrowable` should never be handled by user programs.